Import the most common libraries for Data Science

In [106]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


Read the PREP results of 2024

In [107]:
import os
print(os.getcwd())


/home/jj/Datascience/Springboard/Capstones/Capstone2/INE


In [108]:
import pandas as pd

# Define the correct file path
file_path = './2024_PREP_Presidencia/PRES_2024.csv'

# Try loading the CSV with different delimiters and skipping bad lines
try:
    df = pd.read_csv(file_path, on_bad_lines='skip')
    print("File Loaded Successfully by skipping bad lines")

except Exception as e:
    print("Error reading the file:", e)


File Loaded Successfully by skipping bad lines


/tmp/ipykernel_57839/3485527797.py:8: DtypeWarning: Columns (34,35,36) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, on_bad_lines='skip')


In [109]:
df.head()

CLAVE_CASILLA     CLAVE_ACTA    ID_ENTIDAD     ENTIDAD      \
0  '010000M0100'  '010000M01009'       1      AGUASCALIENTES   
1  '010338B0100'  '010338B01001'       1      AGUASCALIENTES   
2  '010338C0100'  '010338C01001'       1      AGUASCALIENTES   
3  '010338C0200'  '010338C02001'       1      AGUASCALIENTES   
4  '010339B0100'  '010339B01001'       1      AGUASCALIENTES   

   ID_DISTRITO_FEDERAL DISTRITO_FEDERAL  SECCION  ID_CASILLA TIPO_CASILLA  \
0          NaN                    NaN       NaN       NaN          M        
1          1.0            Jesus Maria     338.0       1.0          B        
2          1.0            Jesus Maria     338.0       1.0          C        
3          1.0            Jesus Maria     338.0       2.0          C        
4          1.0            Jesus Maria     339.0       1.0          B        

   EXT_CONTIGUA  ...  CONTABILIZADA MECANISMOS_TRASLADO  \
0        0       ...        1                  D          
1        0       ...        1                F-D          
2        0       ...        1                F-D          
3        0       ...        -                  -          
4        0       ...        1                F-D          

                  CODIGO_INTEGRIDAD                  FECHA_HORA_ACOPIO  \
0  d2a80f2be642eb86ca1227c4172ca106097ad95f7aa035...  03/06/2024 19:17   
1  3a40b592baa0f73bcdfe08789e6f36f8ad6ea1720bbd04...  03/06/2024 02:14   
2  330045156f5e2abe28f938df81ad6beb0db5655fee88f9...  03/06/2024 08:11   
3                                                  -                 -   
4  c76b71b0b172a1a0735e577930c2b1fd911279fee257cd...  03/06/2024 01:27   

  FECHA_HORA_CAPTURA FECHA_HORA_VERIFICACION  ORIGEN  DIGITALIZACION  \
0  03/06/2024 19:17      03/06/2024 19:17        CATD     Escáner      
1  03/06/2024 02:29      03/06/2024 02:29     CASILLA       Móvil      
2  03/06/2024 09:03      03/06/2024 09:04        CATD     Escáner      
3                 -                     -           -           -      
4  03/06/2024 02:34      03/06/2024 02:35     CASILLA       Móvil      

  TIPO_DOCUMENTO COTEJADA  
0    ACTA VMRE       1     
1    Acta PREP       1     
2    Acta PREP       0     
3            -       -     
4    Acta PREP       0     

[5 rows x 48 columns]

In [110]:
df.isnull().sum()


CLAVE_CASILLA                      0
CLAVE_ACTA                         0
ID_ENTIDAD                         0
ENTIDAD                            0
ID_DISTRITO_FEDERAL              466
DISTRITO_FEDERAL                 466
SECCION                          466
ID_CASILLA                       466
TIPO_CASILLA                       0
EXT_CONTIGUA                       0
UBICACION_CASILLA                  0
TIPO_ACTA                        493
TOTAL_BOLETAS_SOBRANTES          466
TOTAL_PERSONAS_VOTARON           466
TOTAL_REP_PARTIDO_CI_VOTARON     466
TOTAL_VOTOS_SACADOS              466
PAN                                0
PRI                                0
PRD                                0
PAN_PRI_PRD                        0
PAN_PRI                            0
PAN_PRD                            0
PRI_PRD                            0
PVEM                               0
PT                                 0
MORENA                             0
PVEM_PT_MORENA                     0
P

In [111]:
df.columns

Index(['CLAVE_CASILLA', 'CLAVE_ACTA', 'ID_ENTIDAD', 'ENTIDAD',
       'ID_DISTRITO_FEDERAL', 'DISTRITO_FEDERAL', 'SECCION', 'ID_CASILLA',
       'TIPO_CASILLA', 'EXT_CONTIGUA', 'UBICACION_CASILLA', 'TIPO_ACTA',
       'TOTAL_BOLETAS_SOBRANTES', 'TOTAL_PERSONAS_VOTARON',
       'TOTAL_REP_PARTIDO_CI_VOTARON', 'TOTAL_VOTOS_SACADOS', 'PAN', 'PRI',
       'PRD', 'PAN_PRI_PRD', 'PAN_PRI', 'PAN_PRD', 'PRI_PRD', 'PVEM', 'PT',
       'MORENA', 'PVEM_PT_MORENA', 'PVEM_PT', 'PVEM_MORENA', 'PT_MORENA', 'MC',
       'NO_REGISTRADAS', 'NULOS', 'TOTAL_VOTOS_ASENTADO',
       'TOTAL_VOTOS_CALCULADO', 'LISTA_NOMINAL', 'REPRESENTANTES_PP_CI',
       'OBSERVACIONES', 'CONTABILIZADA', 'MECANISMOS_TRASLADO',
       'CODIGO_INTEGRIDAD', 'FECHA_HORA_ACOPIO', 'FECHA_HORA_CAPTURA',
       'FECHA_HORA_VERIFICACION', 'ORIGEN', 'DIGITALIZACION', 'TIPO_DOCUMENTO',
       'COTEJADA'],
      dtype='object')

By reading the dictionary of Features, I notice that not all the features (48) are useful for the objective of predicting the vote preference. So in the first filtering process only 26/48 features will be kept:

In [112]:
features_to_keep = [
    'ID_ENTIDAD', 'ENTIDAD', 'ID_DISTRITO_FEDERAL', 'DISTRITO_FEDERAL', 
    'SECCION', 'TIPO_CASILLA', 'UBICACION_CASILLA', 'TOTAL_BOLETAS_SOBRANTES', 
    'TOTAL_PERSONAS_VOTARON', 'PAN', 'PRI', 'PRD', 'PAN_PRI_PRD', 
    'PAN_PRI', 'PAN_PRD', 'PRI_PRD', 'PVEM', 'PT', 'MORENA', 
    'PVEM_PT_MORENA', 'PVEM_PT', 'PVEM_MORENA', 'PT_MORENA', 
    'MC', 'NO_REGISTRADAS', 'NULOS', 'TOTAL_VOTOS_CALCULADO'
]

In [113]:
filtered_df = df[features_to_keep]

In [114]:
# Save the filtered DataFrame to a new CSV file
filtered_csv_file_path = 'filtered_PREP24.csv'
filtered_df.to_csv(filtered_csv_file_path, index=False)

print(f'Filtered DataFrame saved to {filtered_csv_file_path}')

Filtered DataFrame saved to filtered_PREP24.csv


In [115]:
df= filtered_df

With this first Filtering process, reducing # of features from 48 to 26 (by elimination of Non value added columns), the elections report looks more readable.

In [116]:
df.head()

ID_ENTIDAD     ENTIDAD      ID_DISTRITO_FEDERAL DISTRITO_FEDERAL  SECCION  \
0       1      AGUASCALIENTES          NaN                    NaN       NaN    
1       1      AGUASCALIENTES          1.0            Jesus Maria     338.0    
2       1      AGUASCALIENTES          1.0            Jesus Maria     338.0    
3       1      AGUASCALIENTES          1.0            Jesus Maria     338.0    
4       1      AGUASCALIENTES          1.0            Jesus Maria     339.0    

  TIPO_CASILLA  UBICACION_CASILLA TOTAL_BOLETAS_SOBRANTES  \
0       M               1                    NaN            
1       B               2                    263            
2       C               2                    275            
3       C               2                      -            
4       B               2                    261            

  TOTAL_PERSONAS_VOTARON PAN   ...  PT MORENA PVEM_PT_MORENA PVEM_PT  \
0           NaN           874  ...  19   647        157          1     
1           423            69  ...  15   193         10          1     
2           412            74  ...  19   182         13          0     
3             -             -  ...   -     -          -          -     
4           382            64  ...  25   166         15          0     

  PVEM_MORENA PT_MORENA  MC NO_REGISTRADAS NULOS TOTAL_VOTOS_CALCULADO  
0       5         16     64        6         17           2096          
1       1          2     40        0         12            425          
2       0          0     42        0         20            412          
3       -          -      -        -          -              0          
4       1          3     29        0         14            382          

[5 rows x 27 columns]

Let's check again the null values!

In [117]:
df.isnull().sum()


ID_ENTIDAD                   0
ENTIDAD                      0
ID_DISTRITO_FEDERAL        466
DISTRITO_FEDERAL           466
SECCION                    466
TIPO_CASILLA                 0
UBICACION_CASILLA            0
TOTAL_BOLETAS_SOBRANTES    466
TOTAL_PERSONAS_VOTARON     466
PAN                          0
PRI                          0
PRD                          0
PAN_PRI_PRD                  0
PAN_PRI                      0
PAN_PRD                      0
PRI_PRD                      0
PVEM                         0
PT                           0
MORENA                       0
PVEM_PT_MORENA               0
PVEM_PT                      0
PVEM_MORENA                  0
PT_MORENA                    0
MC                           0
NO_REGISTRADAS               0
NULOS                        0
TOTAL_VOTOS_CALCULADO        0
dtype: int64

I see a pattern of 466 empty spaces repeated in the variables:
ID_DISTRITO_FEDERAL        466
DISTRITO_FEDERAL           466
SECCION                    466
TOTAL_BOLETAS_SOBRANTES    466
TOTAL_PERSONAS_VOTARON     466

It is probable that those are null polling stations so I will just need to do a Drop to all of the rows, but first let's double check:

In [118]:
# Filter rows where ID_DISTRITO_FEDERAL is empty
empty_rows = df[df['ID_DISTRITO_FEDERAL'].isnull()]

# Check if remaining specified columns are also empty
columns_to_check = ['ID_DISTRITO_FEDERAL', 'DISTRITO_FEDERAL', 'SECCION', 
                    'TOTAL_BOLETAS_SOBRANTES', 'TOTAL_PERSONAS_VOTARON']

# Display rows where any of these columns are empty
empty_check = empty_rows[columns_to_check].isnull().all(axis=1)

# Display the result
empty_rows_filtered = empty_rows[empty_check]
print(empty_rows_filtered)


        ID_ENTIDAD     ENTIDAD      ID_DISTRITO_FEDERAL DISTRITO_FEDERAL  \
0            1      AGUASCALIENTES         NaN                 NaN         
306          1      AGUASCALIENTES         NaN                 NaN         
795          1      AGUASCALIENTES         NaN                 NaN         
1209         1      AGUASCALIENTES         NaN                 NaN         
1212         1      AGUASCALIENTES         NaN                 NaN         
...            ...             ...                  ...              ...   
168165      32           ZACATECAS         NaN                 NaN         
168860      32           ZACATECAS         NaN                 NaN         
170018      32           ZACATECAS         NaN                 NaN         
170022      32           ZACATECAS         NaN                 NaN         
170600      32           ZACATECAS         NaN                 NaN         

        SECCION TIPO_CASILLA  UBICACION_CASILLA TOTAL_BOLETAS_SOBRANTES  \
0        NaN

Correct, 466 of them are coincident in the same empty values. Now let's drop those 466 rows and rewrite the df

In [119]:

# Once confirmed, drop the empty rows
df = df.dropna(subset=['ID_DISTRITO_FEDERAL', 'DISTRITO_FEDERAL', 'SECCION', 
                               'TOTAL_BOLETAS_SOBRANTES', 'TOTAL_PERSONAS_VOTARON'])

Confirm if all empty rows where eliminated

In [120]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 170944 entries, 1 to 171409
Data columns (total 27 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   ID_ENTIDAD               170944 non-null  int64  
 1   ENTIDAD                  170944 non-null  object 
 2   ID_DISTRITO_FEDERAL      170944 non-null  float64
 3   DISTRITO_FEDERAL         170944 non-null  object 
 4   SECCION                  170944 non-null  float64
 5   TIPO_CASILLA             170944 non-null  object 
 6   UBICACION_CASILLA        170944 non-null  int64  
 7   TOTAL_BOLETAS_SOBRANTES  170944 non-null  object 
 8   TOTAL_PERSONAS_VOTARON   170944 non-null  object 
 9   PAN                      170944 non-null  object 
 10  PRI                      170944 non-null  object 
 11  PRD                      170944 non-null  object 
 12  PAN_PRI_PRD              170944 non-null  object 
 13  PAN_PRI                  170944 non-null  object 
 14  PAN_PRD  

In [121]:
# List of columns to convert to float
columns_to_convert = [
    'TOTAL_BOLETAS_SOBRANTES', 'TOTAL_PERSONAS_VOTARON', 'UBICACION_CASILLA',
    'NO_REGISTRADAS', 'NULOS', 'TOTAL_VOTOS_CALCULADO',
    'PAN', 'PRI', 'PRD', 'PAN_PRI_PRD', 'PAN_PRI', 'PAN_PRD', 'PRI_PRD', 
    'PVEM', 'PT', 'MORENA', 'PVEM_PT_MORENA', 'PVEM_PT', 'PVEM_MORENA', 
    'PT_MORENA', 'MC'
]

# Convert columns to numeric, forcing errors to NaN to handle incorrect values
for col in columns_to_convert:
    df.loc[:, col] = pd.to_numeric(df[col], errors='coerce').astype(float)

In [122]:
df.shape

(170944, 27)

In [123]:
df.isnull().sum()

ID_ENTIDAD                     0
ENTIDAD                        0
ID_DISTRITO_FEDERAL            0
DISTRITO_FEDERAL               0
SECCION                        0
TIPO_CASILLA                   0
UBICACION_CASILLA              0
TOTAL_BOLETAS_SOBRANTES    19931
TOTAL_PERSONAS_VOTARON     18893
PAN                        12660
PRI                        12517
PRD                        13865
PAN_PRI_PRD                17643
PAN_PRI                    21809
PAN_PRD                    25272
PRI_PRD                    25489
PVEM                       12639
PT                         12708
MORENA                     12382
PVEM_PT_MORENA             16216
PVEM_PT                    21553
PVEM_MORENA                19842
PT_MORENA                  19890
MC                         12114
NO_REGISTRADAS             28187
NULOS                      13747
TOTAL_VOTOS_CALCULADO        762
dtype: int64

There is an amazing number of NaN values in the polls
28,187 out of 170,944.
As NaN values represent missing votes, filling them with 0 might be the most logical choice. So the NaN values will be captured as "CERO"

In [124]:
df = df.drop(columns=['TOTAL_VOTOS_CALCULADO'])

In [125]:
df.isnull().sum()

ID_ENTIDAD                     0
ENTIDAD                        0
ID_DISTRITO_FEDERAL            0
DISTRITO_FEDERAL               0
SECCION                        0
TIPO_CASILLA                   0
UBICACION_CASILLA              0
TOTAL_BOLETAS_SOBRANTES    19931
TOTAL_PERSONAS_VOTARON     18893
PAN                        12660
PRI                        12517
PRD                        13865
PAN_PRI_PRD                17643
PAN_PRI                    21809
PAN_PRD                    25272
PRI_PRD                    25489
PVEM                       12639
PT                         12708
MORENA                     12382
PVEM_PT_MORENA             16216
PVEM_PT                    21553
PVEM_MORENA                19842
PT_MORENA                  19890
MC                         12114
NO_REGISTRADAS             28187
NULOS                      13747
dtype: int64

In [126]:
# Fill NaN values with 0
df = df.fillna(0)

/tmp/ipykernel_57839/1326670806.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df = df.fillna(0)


Now is time to clean again the number of columns. I know from PRES_CANDIDATURAS_2024.csv that even when there are 16 'virtual' political parties, there are only 3 politicians competing: Claudia, Xochitl and Maynez.

In [127]:
import pandas as pd

# Define the correct file path
file_path = './2024_PREP_Presidencia/PRES_CANDIDATURAS_2024.csv'

# Try loading the CSV with different delimiters and skipping bad lines
try:
    presidents = pd.read_csv(file_path, on_bad_lines='skip')
    print("File Loaded Successfully by skipping bad lines")

except Exception as e:
    print("Error reading the file:", e)

File Loaded Successfully by skipping bad lines


In [128]:
print('this is the table of presidents splited by political party:')
presidents.head(16)

this is the table of presidents splited by political party:


PARTIDO_CI   ID_ENTIDAD ID_DISTRITO_FEDERAL   CANDIDATURA_PROPIETARIA  
0              PAN  Nacional        Nacional       Bertha Xóchitl Gálvez Ruíz
1              PRI  Nacional        Nacional       Bertha Xóchitl Gálvez Ruíz
2              PRD  Nacional        Nacional       Bertha Xóchitl Gálvez Ruíz
3             PVEM  Nacional        Nacional          Claudia Sheinbaum Pardo
4               PT  Nacional        Nacional          Claudia Sheinbaum Pardo
5               MC  Nacional        Nacional             Jorge Álvarez Máynez
6           MORENA  Nacional        Nacional          Claudia Sheinbaum Pardo
7      PAN-PRI-PRD  Nacional        Nacional       Bertha Xóchitl Gálvez Ruíz
8          PAN-PRI  Nacional        Nacional       Bertha Xóchitl Gálvez Ruíz
9          PAN-PRD  Nacional        Nacional       Bertha Xóchitl Gálvez Ruíz
10         PRI-PRD  Nacional        Nacional       Bertha Xóchitl Gálvez Ruíz
11  PVEM_PT_MORENA  Nacional        Nacional          Claudia Sheinbaum Pardo
12         PVEM_PT  Nacional        Nacional          Claudia Sheinbaum Pardo
13     PVEM_MORENA  Nacional        Nacional          Claudia Sheinbaum Pardo
14       PT_MORENA  Nacional        Nacional          Claudia Sheinbaum Pardo

So I will drop the 16 columns and add all the votes in only 3 columns

In [129]:

# Sum the votes for each candidate
df['Xochitl'] = df[['PAN', 'PRI', 'PRD', 'PAN_PRI_PRD', 'PAN_PRI', 'PAN_PRD', 'PRI_PRD']].sum(axis=1)
df['Claudia'] = df[['PVEM', 'PT', 'MORENA', 'PVEM_PT_MORENA', 'PVEM_PT', 'PVEM_MORENA', 'PT_MORENA']].sum(axis=1)
df['Maynez'] = df['MC']

# List of columns to drop
columns_to_drop = [
    'PAN', 'PRI', 'PRD', 'PAN_PRI_PRD', 'PAN_PRI', 'PAN_PRD', 'PRI_PRD', 
    'PVEM', 'PT', 'MORENA', 'PVEM_PT_MORENA', 'PVEM_PT', 'PVEM_MORENA', 
    'PT_MORENA', 'MC'
]

# Drop the columns
df = df.drop(columns=columns_to_drop)


In [130]:
df.head()

ID_ENTIDAD     ENTIDAD      ID_DISTRITO_FEDERAL DISTRITO_FEDERAL  SECCION  \
1       1      AGUASCALIENTES          1.0            Jesus Maria     338.0    
2       1      AGUASCALIENTES          1.0            Jesus Maria     338.0    
3       1      AGUASCALIENTES          1.0            Jesus Maria     338.0    
4       1      AGUASCALIENTES          1.0            Jesus Maria     339.0    
5       1      AGUASCALIENTES          1.0            Jesus Maria     339.0    

  TIPO_CASILLA  UBICACION_CASILLA  TOTAL_BOLETAS_SOBRANTES  \
1       B               2                   263.0            
2       C               2                   275.0            
3       C               2                     0.0            
4       B               2                   261.0            
5       C               2                   266.0            

   TOTAL_PERSONAS_VOTARON  NO_REGISTRADAS  NULOS  Xochitl  Claudia  Maynez  
1           423.0                0.0       12.0    134.0    239.0    40.0   
2           412.0                0.0       20.0    119.0    231.0    42.0   
3             0.0                0.0        0.0      0.0      0.0     0.0   
4           382.0                0.0       14.0    114.0    225.0    29.0   
5           377.0                0.0        9.0    116.0    219.0    28.0

In [131]:
# Summarize the votes per candidate
summary = df[['Xochitl', 'Claudia', 'Maynez']].sum()

# Calculate the total votes across all candidates
total_votes = summary.sum()

# Calculate the percentage of votes per candidate
percentage = (summary / total_votes) * 100

# Combine the summary and percentage into a single DataFrame
summary_df = pd.DataFrame({
    'Total Votes': summary,
    'Percentage': percentage
})

print(summary_df)

         Total Votes  Percentage
Xochitl  15621292.0    28.517491
Claudia  33305398.0    60.800758
Maynez    5851242.0    10.681750


In [132]:

# Group by state and summarize the votes per candidate
grouped = df.groupby('ENTIDAD')[['Xochitl', 'Claudia', 'Maynez']].sum()

# Calculate the total votes across all candidates for each state
grouped['Total Votes'] = grouped.sum(axis=1)

# Calculate the percentage of votes per candidate for each state
grouped['X %'] = (grouped['Xochitl'] / grouped['Total Votes']) * 100
grouped['C %'] = (grouped['Claudia'] / grouped['Total Votes']) * 100
grouped['M %'] = (grouped['Maynez'] / grouped['Total Votes']) * 100

# Round the percentage columns to 2 decimal places
grouped = grouped.round({'X %': 2, 'C %': 2, 'M %': 2})

# Display the results
print(grouped)

                      Xochitl    Claudia    Maynez   Total Votes   X %   \
ENTIDAD                                                                   
AGUASCALIENTES        294746.0   270389.0   57054.0    622189.0   47.37   
BAJA CALIFORNIA       290698.0   862661.0  132914.0   1286273.0   22.60   
BAJA CALIFORNIA SUR    90838.0   175755.0   26581.0    293174.0   30.98   
CAMPECHE               62268.0   240693.0   82898.0    385859.0   16.14   
CHIAPAS               355592.0  1519559.0  162166.0   2037317.0   17.45   
CHIHUAHUA             495324.0   744823.0  135550.0   1375697.0   36.01   
CIUDAD DE MEXICO     1862780.0  2978243.0  454084.0   5295107.0   35.18   
COAHUILA              595270.0   813432.0   84184.0   1492886.0   39.87   
COLIMA                102288.0   182063.0   45426.0    329777.0   31.02   
DURANGO               227617.0   402566.0   51520.0    681703.0   33.39   
GUANAJUATO           1057455.0  1223770.0  278420.0   2559645.0   41.31   
GUERRERO              276

In [134]:
grouped = grouped.reset_index()
# Display the DataFrame with improved formatting
pd.set_option('display.colheader_justify', 'center')
grouped.head(32)

index       ENTIDAD          Xochitl    Claudia    Maynez   Total Votes  \
0     0         AGUASCALIENTES   294746.0   270389.0   57054.0    622189.0    
1     1        BAJA CALIFORNIA   290698.0   862661.0  132914.0   1286273.0    
2     2    BAJA CALIFORNIA SUR    90838.0   175755.0   26581.0    293174.0    
3     3               CAMPECHE    62268.0   240693.0   82898.0    385859.0    
4     4                CHIAPAS   355592.0  1519559.0  162166.0   2037317.0    
5     5              CHIHUAHUA   495324.0   744823.0  135550.0   1375697.0    
6     6       CIUDAD DE MEXICO  1862780.0  2978243.0  454084.0   5295107.0    
7     7               COAHUILA   595270.0   813432.0   84184.0   1492886.0    
8     8                 COLIMA   102288.0   182063.0   45426.0    329777.0    
9     9                DURANGO   227617.0   402566.0   51520.0    681703.0    
10   10             GUANAJUATO  1057455.0  1223770.0  278420.0   2559645.0    
11   11               GUERRERO   276953.0  1039283.0   95579.0   1411815.0    
12   12                HIDALGO   299831.0  1001963.0  156217.0   1458011.0    
13   13                JALISCO  1303587.0  1592942.0  631761.0   3528290.0    
14   14                 MEXICO  2177817.0  4916419.0  876926.0   7971162.0    
15   15              MICHOACAN   562486.0  1033704.0  232501.0   1828691.0    
16   16                MORELOS   195923.0   578230.0  117582.0    891735.0    
17   17                NAYARIT   102576.0   306423.0   67984.0    476983.0    
18   18             NUEVO LEON   839701.0  1062016.0  432115.0   2333832.0    
19   19                 OAXACA   261778.0  1367731.0  119132.0   1748641.0    
20   20                 PUEBLA   694569.0  2066399.0  322725.0   3083693.0    
21   21              QUERETARO   428194.0   585662.0  109805.0   1123661.0    
22   22           QUINTANA ROO   129343.0   536134.0   57373.0    722850.0    
23   23        SAN LUIS POTOSI   330455.0   773086.0  138705.0   1242246.0    
24   24                SINALOA   301345.0   782221.0   95577.0   1179143.0    
25   25                 SONORA   270687.0   674746.0   99032.0   1044465.0    
26   26                TABASCO   115104.0   811270.0   61696.0    988070.0    
27   27             TAMAULIPAS   390288.0   910332.0  139605.0   1440225.0    
28   28               TLAXCALA   106484.0   486946.0   98955.0    692385.0    
29   29               VERACRUZ   810805.0  2300252.0  281570.0   3392627.0    
30   30                YUCATAN   365357.0   696450.0   81057.0   1142864.0    
31   31              ZACATECAS   223133.0   369235.0  124548.0    716916.0    

     X %    C %    M %   
0   47.37  43.46   9.17  
1   22.60  67.07  10.33  
2   30.98  59.95   9.07  
3   16.14  62.38  21.48  
4   17.45  74.59   7.96  
5   36.01  54.14   9.85  
6   35.18  56.25   8.58  
7   39.87  54.49   5.64  
8   31.02  55.21  13.77  
9   33.39  59.05   7.56  
10  41.31  47.81  10.88  
11  19.62  73.61   6.77  
12  20.56  68.72  10.71  
13  36.95  45.15  17.91  
14  27.32  61.68  11.00  
15  30.76  56.53  12.71  
16  21.97  64.84  13.19  
17  21.51  64.24  14.25  
18  35.98  45.51  18.52  
19  14.97  78.22   6.81  
20  22.52  67.01  10.47  
21  38.11  52.12   9.77  
22  17.89  74.17   7.94  
23  26.60  62.23  11.17  
24  25.56  66.34   8.11  
25  25.92  64.60   9.48  
26  11.65  82.11   6.24  
27  27.10  63.21   9.69  
28  15.38  70.33  14.29  
29  23.90  67.80   8.30  
30  31.97  60.94   7.09  
31  31.12  51.50  17.37

In [136]:
# Sort the DataFrame by 'Claudia %' in ascending order
sorted_df = grouped.sort_values(by='C %')

# Display the sorted DataFrame
sorted_df.head(10)

index      ENTIDAD        Xochitl    Claudia    Maynez   Total Votes  \
0     0      AGUASCALIENTES   294746.0   270389.0   57054.0    622189.0    
13   13             JALISCO  1303587.0  1592942.0  631761.0   3528290.0    
18   18          NUEVO LEON   839701.0  1062016.0  432115.0   2333832.0    
10   10          GUANAJUATO  1057455.0  1223770.0  278420.0   2559645.0    
31   31           ZACATECAS   223133.0   369235.0  124548.0    716916.0    
21   21           QUERETARO   428194.0   585662.0  109805.0   1123661.0    
5     5           CHIHUAHUA   495324.0   744823.0  135550.0   1375697.0    
7     7            COAHUILA   595270.0   813432.0   84184.0   1492886.0    
8     8              COLIMA   102288.0   182063.0   45426.0    329777.0    
6     6    CIUDAD DE MEXICO  1862780.0  2978243.0  454084.0   5295107.0    

     X %    C %    M %   
0   47.37  43.46   9.17  
13  36.95  45.15  17.91  
18  35.98  45.51  18.52  
10  41.31  47.81  10.88  
31  31.12  51.50  17.37  
21  38.11  52.12   9.77  
5   36.01  54.14   9.85  
7   39.87  54.49   5.64  
8   31.02  55.21  13.77  
6   35.18  56.25   8.58

In [137]:
grouped.to_csv('grouped_by_state.csv', index=False)